### PR TITLE
bpo-17140: Document multiprocessing's ThreadPool

### PR DESCRIPTION
**Pull request by godlygeek**
Up until now the `multiprocessing.pool.ThreadPool` class has gone
undocumented, despite being a public class in multiprocessing that is
included in `multiprocessing.pool.__all__`.

https://bugs.python.org/issue17140

### CODE REVIEW

**Review by Fidget-Spinner — APPROVED** [review]
This LGTM. The only comment I have is that it seems strange to place `ThreadPool` under the `Process Pools` section, though I understand that it makes sense since it's a subclass of `Pool`. Maybe we can create a `Thread Pools` subsection for it. What do you think?

**Comment by godlygeek** [comment]
I almost did that, and don't mind changing it to be that way if you prefer. Would you like the new section directly below the "Process Pools" section?

I also considered putting it in the `The :mod:`multiprocessing.dummy` module` section, which is slightly weird since it's in `multiprocessing.pool`, but also is arguably the place where it best belongs since it is related to the "multiprocessing but with threads instead of processes" thing that `multiprocessing.dummy` does, and because `multiprocessing.dummy.Pool` does return `ThreadPool` instances (though it is a wrapper function around `multiprocessing.Pool.ThreadPool` for some reason, not an alias for it).

**Comment by Fidget-Spinner** [comment]
> I almost did that, and don't mind changing it to be that way if you prefer. Would you like the new section directly below the "Process Pools" section?

Hmm I think, I'll wait and see what the core dev who reviews this prefers. Thanks for working on this :).

File: Doc/library/multiprocessing.rst
Comment (pablogsal) [inline Doc/library/multiprocessing.rst]: Two minor suggestions:

* Add a reference to this class in the `multiprocessing.dummy` section.
* Add a reference to the `concurrent.futures` module or thread-specific-APIs as these are more robust APIs in general for what this class offers and unfortunatelly not that many people know about them.

Comment (godlygeek) [reply Doc/library/multiprocessing.rst]: > Add a reference to this class in the multiprocessing.dummy section.

Do you think that this should
1. move to the `multiprocessing.dummy` section in its entirety,
2. stay in the "Process Pools" section and be linked to from the `multiprocessing.dummy` section, or
3. go into a new "Thread Pools" section (and if so, where should the new section be placed in the document)?

> Add a reference to the `concurrent.futures` module

Seems reasonable to make a suggestion that guides people towards `concurrent.futures`, but that module currently isn't referenced by the `multiprocessing` docs at all. Are you thinking that should be scoped to just `multiprocessing.Pool.ThreadPool` vs `concurrent.futures.ThreadPoolExecutor`, or should it also mention `multiprocessing.Pool` vs `concurrent.futures.ProcessPoolExecutor`?

Perhaps let's refine the wording here before I update the PR. Are you thinking something like:

> Most users should instead prefer to use a `concurrent.futures.Executor`, which provides a simpler interface to thread pools and process pools than the one provided here. These are usually easier to use correctly, and other modules (like `asyncio`) have builtin support for handling the `concurrent.futures.Future` instances that these pools return for tracking the progress of jobs that are run asynchronously.

Is that about it, or are there more specific points I should make?

> or thread-specific-APIs

I'm not sure what you mean by this bit.

Comment (pablogsal) [reply Doc/library/multiprocessing.rst]: > Do you think that this should

I don't mind that much if is on `multiprocessing.dummy` or if `multiprocessing.dummy` refers to this but I would prefer option 1. The reason is that I don't think is a good idea to have a "Thread Pools" section because this is still the multiprocessing module and not only the APIs have been designed without threads initially in mind but also is slightly confusing to users (why is not on "threading"?).

> Are you thinking that should be scoped to just multiprocessing.Pool.ThreadPool vs concurrent.futures.ThreadPoolExecutor

Just scoped for `multiprocessing.Pool.ThreadPool` vs `concurrent.futures.ThreadPoolExecutor`. The reason is that `multiprocessing.Pool.ThreadPool` doesn't have the most friendly API and suffers from some of the design that is process-specific. It also doesn't share some of the functionality (like using `maxtasksperchild`). IIRC, this was partially why this class was never fully documented.

> I'm not sure what you mean by this bit.

Sorry, that was poorly written. I meant to reference https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor.

**Comment by godlygeek** [comment]
@pablogsal OK, I've force-pushed a new commit that yanks it down into the `multiprocessing.dummy` section, and explains that `ThreadPoolExecutor` is generally preferable. Take a look and let me know if you think there's anything else it needs, please.

**Review by pablogsal — APPROVED** [review]
LGTM

I will leave some days in case Antoine has time to review this before landing in case he has some new insights or he wants some changes.

**Comment by godlygeek** [comment]
I just made one final change - I had the note about preferring `ThreadPoolExecutor` above the class definition, and I've moved it down into the documentation for the class instead.

**Comment by godlygeek** [comment]
One more force push - I noticed a missing word after re-reading this.

**Comment by pablogsal** [comment]
> One more force push - I noticed a missing word after re-reading this.

Note the there is no need for force pushes: the commits will be squashed and the core Dev merging will write the final commit anyway.

**Comment by pitrou** [comment]
This is fine with me. Thanks for reviewing @pablogsal !

**Pull request merged**
